include/glow/Support/Log.h

```diff
@@ -0,0 +1,301 @@
+/**
+ * Copyright (c) 2018-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_SUPPORT_LOG_H
+#define GLOW_SUPPORT_LOG_H
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <fstream>
+#include <string>
+
+namespace glow {
+
+extern llvm::cl::opt<bool> dumpCompilationLogOpt;
+
+/// A class for logging all compilation related activities.
+class LogContext {
+private:
+  /// A vector that keeps track of all log contents.
+  std::vector<std::string> logContents_;
+
+  /// A vector that keeps track of current compilation scopes.
+  std::vector<std::string> logScopes_;
+
+  /// A string that represents the current full scope name.
+  std::string currentFullScope_ = "";
+
+public:
+  LogContext(){};
+
+  /// Add content into the contents vector.
+  void addLogContent(llvm::StringRef logContent);
+
+  /// Add a scope into the scope vector.
+  void addLogScope(llvm::StringRef scopeName);
+
+  /// Getter that returns the current full scope name.
+  llvm::StringRef getFullScopeName() { return currentFullScope_; }
+
+  /// Pops out the most recently added scope.
+  void popBackLogScope();
+
+  /// Dumps the log into a file named after the given \p function name.
+  void dumpLog(llvm::StringRef funcName);
+};
+
+/// Logs the node creation with no input node. (Example: Splat)
+#define LOG_NODE_CREATE(newNode)                                               \
```

Comment:
Does it need to be a macro? Or can it be a function?

Reply:
If it can be a function, it could be declared here, but implemented in a cpp file.

Reply:
put into a function now. But i can't put that in Log.h or Log.cpp as I can't include the Node.h in that scope, which may cause include errors. So i put the function definition in Graph.h